File: Intatis-Windows/src/Intatis.Core/Ids.cs
using System.Security.Cryptography;

namespace Intatis.Core;

public enum SessionKind
{
    Chat,
    Code,
    Cowork,
}

public static class SessionKindExtensions
{
    public static string ToWire(this SessionKind kind) => kind switch
    {
        SessionKind.Chat => "chat",
        SessionKind.Code => "code",
        SessionKind.Cowork => "cowork",
        _ => "chat",
    };

    public static SessionKind FromWire(string value) => value switch
    {
        "code" => SessionKind.Code,
        "cowork" => SessionKind.Cowork,
        _ => SessionKind.Chat,
    };

    /// <summary>The session ID prefix also encodes the kind (sess_ / code_ / cowork_).</summary>
    public static string IdPrefix(this SessionKind kind) => kind switch
    {
        SessionKind.Chat => "sess_",
        SessionKind.Code => "code_",
        SessionKind.Cowork => "cowork_",
        _ => "sess_",
    };

    public static bool UsesWorkspace(this SessionKind kind) => kind != SessionKind.Chat;
}

/// <summary>
/// Typed identifiers mirror the Apple project's TypedID scheme: random
/// lowercase alphanumerics of length 8 behind a short prefix such as
/// "sess_" or "msg_". They serialize as bare JSON strings.
/// </summary>
public readonly record struct TypedId(string Value)
{
    public override string ToString() => Value;
    public static implicit operator string(TypedId id) => id.Value;
}

public static class IdGen
{
    private const string Alphabet = "abcdefghijklmnopqrstuvwxyz0123456789";

    public static string Random(string prefix, int length = 8)
    {
        var bytes = RandomNumberGenerator.GetBytes(length);
        var chars = new char[length];
        for (var i = 0; i < length; i++)
            chars[i] = Alphabet[bytes[i] % Alphabet.Length];
        return prefix + new string(chars);
    }
}

public readonly record struct SessionId(string Value)
{
    public static SessionId New(SessionKind kind) => new(IdGen.Random(kind.IdPrefix()));
    public SessionKind Kind => Value.StartsWith("code_") ? SessionKind.Code
        : Value.StartsWith("cowork_") ? SessionKind.Cowork
        : SessionKind.Chat;
    public static implicit operator string(SessionId id) => id.Value;
}

public readonly record struct MessageId(string Value)
{
    public static MessageId New() => new(IdGen.Random("msg_"));
    public static implicit operator string(MessageId id) => id.Value;
}

public readonly record struct SubmissionId(string Value)
{
    public static SubmissionId New() => new(IdGen.Random("sub_"));
    public static implicit operator string(SubmissionId id) => id.Value;
}

public readonly record struct TurnId(string Value)
{
    public static TurnId New() => new(IdGen.Random("turn_"));
    public static implicit operator string(TurnId id) => id.Value;
}

public readonly record struct TaskId(string Value)
{
    public static TaskId New() => new(IdGen.Random("task_"));
    public static implicit operator string(TaskId id) => id.Value;
}

public readonly record struct WorkTaskId(string Value)
{
    public static WorkTaskId New() => new(IdGen.Random("wt_"));
    public static implicit operator string(WorkTaskId id) => id.Value;
}

public readonly record struct GoalId(string Value)
{
    public static GoalId New() => new(IdGen.Random("goal_"));
    public static implicit operator string(GoalId id) => id.Value;
}

public readonly record struct ArtifactId(string Value)
{
    public static ArtifactId New() => new(IdGen.Random("art_"));
    public static implicit operator string(ArtifactId id) => id.Value;
}

public readonly record struct RequestId(string Value)
{
    public static RequestId New() => new(IdGen.Random("req_"));
    public static implicit operator string(RequestId id) => id.Value;
}

File: Intatis-Android/shared/src/main/kotlin/com/intatis/shared/Ids.kt
package com.intatis.shared

import java.security.SecureRandom

enum class SessionKind(val wire: String) {
    CHAT("chat"),
    CODE("code"),
    COWORK("cowork");

    val idPrefix: String
        get() = when (this) {
            CHAT -> "sess_"
            CODE -> "code_"
            COWORK -> "cowork_"
        }

    companion object {
        fun fromWire(value: String): SessionKind = when (value) {
            "code" -> CODE
            "cowork" -> COWORK
            else -> CHAT
        }
    }
}

/**
 * Typed identifiers mirror the Apple project's TypedID scheme: random lowercase
 * alphanumerics of length 8 behind a short prefix such as "sess_" or "msg_".
 */
@JvmInline
value class SessionId(val value: String) {
    val kind: SessionKind
        get() = when {
            value.startsWith("code_") -> SessionKind.CODE
            value.startsWith("cowork_") -> SessionKind.COWORK
            else -> SessionKind.CHAT
        }

    companion object {
        fun new(kind: SessionKind) = SessionId(IdGen.random(kind.idPrefix))
    }
}

@JvmInline value class MessageId(val value: String) {
    companion object { fun new() = MessageId(IdGen.random("msg_")) }
}

@JvmInline value class SubmissionId(val value: String) {
    companion object { fun new() = SubmissionId(IdGen.random("sub_")) }
}

@JvmInline value class TurnId(val value: String) {
    companion object { fun new() = TurnId(IdGen.random("turn_")) }
}

@JvmInline value class ArtifactId(val value: String) {
    companion object { fun new() = ArtifactId(IdGen.random("art_")) }
}

object IdGen {
    private const val Alphabet = "abcdefghijklmnopqrstuvwxyz0123456789"
    private val random = SecureRandom()

    fun random(prefix: String, length: Int = 8): String {
        val bytes = ByteArray(length)
        random.nextBytes(bytes)
        return prefix + bytes.joinToString("") { Alphabet[it.toInt().mod(Alphabet.length)].toString() }
    }
}
